query-algebrizer/src/clauses/mod.rs

```diff
@@ -873,6 +968,26 @@ impl ConjoiningClauses {
 }
 
 impl ConjoiningClauses {
+    pub fn apply_clauses(&mut self, schema: &Schema, where_clauses: Vec<WhereClause>) -> Result<()> {
+        let mut deferred = Vec::with_capacity(where_clauses.len());
+        // We apply (top level) type predicates first as an optimization.
+        for c in where_clauses {
```

Comment:
If you're going to collect into `deferred`, then use `into_iter` here to consume the input `Vec`… and then you can even use a filter and `collect` to do it.

But I have a better suggestion. Can we just add a `WhereClause::TypeAnnotation` case to make your check efficient, and go back to walking once by reference and once by value? Maybe a `good-first-bug` follow-up, 'cos it adds a bit more work to something that already works.

Reply:
I just went ahead and did this because it bothered me less than the alternative. It also didn't take nearly as long as (for example) the translator changes.